src/globus_sdk/client.py

```diff
@@ -176,20 +174,89 @@ def _resolve_base_url(cls, init_base_url: str | None, environment: str) -> str:
+        # "In the face of ambiguity, refuse the temptation to guess."
         if self._app:
-            if self.resource_server is None:
-                raise ValueError(
-                    "Unable to use an 'app' with a client with no "
-                    "'resource_server' defined."
-                )
+            raise exc.GlobusSDKUsageError(
+                f"Cannot attach GlobusApp to {type(self).__name__} when one is "
+                "already attached."
+            )
+        if self.app_scopes:
+            # technically, we *could* allow for this, but it's not clear what
+            # it would mean if a user wrote the following:
+            #
+            #   c = ClientClass()
+            #   c.app_scopes = [foo]
+            #   c.attach_globus_app(app, app_scopes=[bar])
+            #
+            # did the user expect a merge, overwrite, or other behavior?
+            raise exc.GlobusSDKUsageError(
+                f"Cannot attach GlobusApp to {type(self).__name__} when `app_scopes` "
+                "is already set. "
+                "The scopes for this client cannot be consistently resolved."
+            )
+        if self.authorizer:
+            raise exc.GlobusSDKUsageError(
+                f"Cannot attach GlobusApp to {type(self).__name__} when it "
+                "has an authorizer assigned."
+            )
+        if self.resource_server is None:
+            raise exc.GlobusSDKUsageError(
+                "Unable to use an 'app' with a client with no "
+                "'resource_server' defined."
+            )
+        # the client's environment must match the app's -- we can't tell which of these
+        # were set explicitly programmatically and which were implicit (e.g., from the
+        # env var), so erroring is the safest option
+        #
+        # although we can track the value used in init (e.g., add
+        # `self._init_environment = environment` to init), it's harder to keep track of
+        # usages like
+        # `client = FooClient(); client.environment = "test"; client.base_url = ...`
+        if self.environment != app.config.environment:
+            raise exc.GlobusSDKUsageError(
+                f"[Environment Mismatch] {type(self).__name__}'s environment "
+                f"({self.environment}) does not match the GlobusApp's configured "
+                f"environment ({app.config.environment})."
+            )
 
-            if self.app_scopes:
-                scope_requirements = self.app_scopes
-            else:
-                scope_requirements = self.default_scope_requirements
+        # now, assign the app, app_name, and scopes
+        self._app = app
+        self.app_scopes = app_scopes
+        if self.app_name is None:
+            self.app_name = app.app_name
+
+        if self.app_scopes:
+            scope_requirements = self.app_scopes
+        else:
+            scope_requirements = self.default_scope_requirements
```

Comment:
Any benefit to assigning `app_scopes` as a client-level instance attribute comes from exposure of "this is what we passed to the app". In the case that app scopes aren't supplied explicitly, we pass the default scope requirements to the app & should reflect that in the attribute.

```suggestion
        self.app_scopes = app_scopes or self.default_scope_requirements
        if self.app_name is None:
            self.app_name = app.app_name
```

Reply:
I'm a bit reticent to change this, since the prior logic produced a `BaseClient.app_scopes` of `None` in the default case and this would change that. It's a very small change, but I don't want to change the behavior of that attribute without first acknowledging that we would be doing so.

Given that `default_scope_requirements` already exists as an attribute, I think it's better for `self.app_scopes` to be defined as "the scopes which were passed on init or when the app was attached", not "the scopes which were passed to the app itself".

Reply:
Fair callout; this would be a slight behavioral change.

It's quite confusing to me that this would be the case:
```python
app = UserApp(client_id='my-cool-client')
flows = FlowsClient(app=app)

# Error: None != [...whatever the default flows scopes are...]
assert flows.app_scopes == app.scope_requirements[flows.resource_server]
```

So much so that I'm happy to have that slight behavioral shift.

Reply:
I disagree because I _don't_ consider that comparison mismatching to be confusing. Specifically because of these two cases:
```python
# setup
app = UserApp(client_id='my-cool-client')
flows_client = FlowsClient(app=app)

...

# case 1: another client comes along
other_flows_client = FlowsClient(app=app, app_scopes=FlowsClient.scopes.manage_runs)

...

# case 2: explicit scope manipulation
app.add_scope_requirements({flows_client.resource_server: [...]})

...

# boom!
assert flows_client.app_scopes == app.scope_requirements[flows_client.resource_server]
```

If you want to know what scopes an app will use, you need to ask the app itself -- asking the client is unsafe.

I'm finding the argument to change it unconvincing because I don't think there's a clear principle that we'd be following here. So it's not outweighing the my default towards backwards compatibility.

---

One thing that I'm unclear about as I look at this body of code is this:

Why is `BaseClient.app_scopes` an attribute at all? Why are we recording the value which (as of today) is passed on init?

In the case of explicit scope manipulation on the app, it may be recorded with a non-`None` value but inaccurate.

I thought we were just recording data from init-time for potential debug purposes. But you seem to be saying you have some other expectation about what that value means, and that leaves me confused.

Reply:
I'd also be happy to remove the attribute.

From my initial comment:
> Any benefit to assigning app_scopes as a client-level instance attribute comes from exposure of "this is what we passed to the app"

That was very intentionally phrased cause I'm not convinced that there is actually any benefit to assigning app_scopes as a client-level instance attribute.

---

In your rebuttal example, I wouldn't expect `flows_client` to combine all scopes registered on the app into its `app_scopes` attribute. I would expect it to report the portion of the combined scopes that it brought to the party.

If flows_client_1 brings the default `chips & dip` to the party, flows_client_2 brings `soda`, and flows_client_3 brings `dino nuggets` then I expect that:
  * `flows_client_1.what_did_you_bring()` returns `chips & dip`
  * `flows_client_2.what_did_you_bring()` returns `soda`
  * `flows_client_3.what_did_you_bring()` return `dino nuggets`
and ultimately
  * `party_app.what_did_all_the_flows_clients_brings()` returns `[chips & dip, soda, dino nuggets]`

Reply:
Okay, let's save removal for future work. If we're both game to remove it, that's an option to consider.
(I tend slightly towards making it a private attribute which records what was passed, since it could help in a debugging session but isn't otherwise important to record in that place.)

I do find your party_app example pretty delightful. And I think the analogy is strong -- the rationale there is "`app_scopes` records what this client passed to the app".

The only real issue I have is that the result should have been `[chips, dip, soda, dino nuggests]`. Clearly chips and dip are separate items.

Given that `client.app_scopes` is already not super-useful, I'm okay with including the change now that I have a more complete rationale for why we should change it. It's such a marginal incompatibility, it's unlikely to hurt anyone. I will note it in the changelog though.

Reply:
> The only real issue I have is that the result should have been [chips, dip, soda, dino nuggests]. Clearly chips and dip are separate items.

Your feedback is appreciated; will update in a followon PR.

Reply:
I gave this nice, neutral-sounding language in the changelog which doesn't call this a fix or a breaking change. Considering this resolved, please reopen if you feel the need to respond to my `chips, dip` provocation!